geo_tag/geo_tag.py

```diff
@@ -0,0 +1,200 @@
+            self.coord = coord.get('coordinates')
+        else:
+            # 2. get the central point of the bounding_box
+            if place is not None:
+                bounding_box = place.get('bounding_box').get('coordinates')[0]
+                ''' Twitter has some wield format historically, though it still rectangle, 
+                but it is not always in (sw, se, ne, nw) order
+                '''
+                swLog = min(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                swLat = min(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+                neLog = max(bounding_box[0][0], bounding_box[1][0], bounding_box[2][0], bounding_box[3][0])
+                neLat = max(bounding_box[0][1], bounding_box[1][1], bounding_box[2][1], bounding_box[3][1])
+
+                # AsterixDB is unhappy with this kind of point "rectangular"
+                if (swLog == neLog and swLat == neLat):
+                    swLog = neLog - 0.0000001
+                    swLat = neLat - 0.0000001
+
+                if (swLog > neLog or swLat > neLat):
+                    print("[ERROR] invalid coordinates of SW and NE.")
+                    return
+
+                # calculate the central point in the rectangle
+                self.coord = ((swLog+neLog) / 2, (swLat+neLat) / 2)
+
+    # help function
+    def extract_geo_tag_from_city_and_state(self, city_state_name, flag):
+        if city_state_name is None:
+            print("[ERROR] There is no city_name.")
+            return
+
+        if "," not in city_state_name:
+            print("[ERROR] parse the location name.")
+            return
+
+        state_of_city = city_state_name.split(",")[1]
+        target_state = str(state_of_city).strip()
+        full_state_name = self.abbrev_us_state[target_state]
+
+        # search it in the cache(hashmap) to get the county's name
+        target_key = city_state_name.split(",")[0] + ', ' + full_state_name
+
+        if not self.coord:
+            # self.coord is empty now
+            geo_locator = Nominatim()
+            location = geo_locator.geocode(target_key)
+            self.coord = (location.longitude, location.latitude)
+
+        geo_content = self.cache.get(target_key)
+        if geo_content is not None:
+            geo_tuple = dict()
+            geo_tuple["stateID"] = geo_content[2]
+            geo_tuple["stateName"] = full_state_name
+            geo_tuple["countyID"] = geo_content[1]
+            geo_tuple["countyName"] = geo_content[3]
+            geo_tuple["cityID"] = geo_content[0]
+            geo_tuple["cityName"] = city_state_name.split(",")[0]
+            geo_tuple["coordinate"] = self.coord
+            geo_tuple["source"] = flag
+
+            self.data['geo_tag'] = geo_tuple
+            print(geo_tuple)
+            
+        else:
+            print("[ERROR] Not find the county of", target_key)
+
+    def infer_geo_from_place(self):
```

Comment:
try not to pass all information through `self`, it makes the code hard to read. you may want change the API of this method.

Reply:
Since `self` contains many variables, it is very convenient to use them. Any suggestions?